ttnn/cpp/ttnn/operations/core.hpp

```diff
@@ -152,58 +155,309 @@ inline ttnn::Tensor squeeze_from_4D(const ttnn::Tensor& tensor, const int rank)
     }
 }
 
-inline ttnn::Tensor to_memory_config(
-    const ttnn::Tensor& tensor,
-    const ttnn::MemoryConfig& memory_config,
-    std::optional<ttnn::DataType> dtype = std::nullopt) {
-    const auto original_memory_config = ttnn::get_memory_config(tensor);
-    if (original_memory_config.has_value() && original_memory_config.value() == memory_config) {
-        return tensor;
+class OperationMonad;
+
+template <class ConcreteOperation, class... Args>
+constexpr auto is_an_invocable_monad = []() -> bool {
+    return is_invocable_v<ConcreteOperation, OperationMonad&, Args...>;
+    // TODO: and std::is_same_v<std::invoke_result_t<ConcreteOperation>, OperationMonad>;
+};
+
+template <typename T>
+constexpr void check_const() {
+    static_assert(std::is_const_v<std::remove_reference_t<T>>, "Argument must be const");
+}
+
+template <typename... Args>
+constexpr void enforce_const_args(Args&&... args) {
+    (..., (check_const<Args>(), void()));
+}
+
+class OperationMonad {
+    const OptionalTensors tensors;
+
+    static void launch_async(
```

Comment:
Is this function meant as a replacement for `launch_op`? I'm not entirely sure what purpose this serves at the ttnn level. I'm wondering if it makes sense to move this down to where the ops are defined and use the Monad Pattern there. 

Additionally, this function has an assert that requires the output tensors to have workers (this means that the tensors are on device). I see that in the `bind` call we support host operations (which produce tensors on host). Is there some special handling here that I'm missing?

Reply:
Yes, we are essentially replacing launch_op.   There are two cases here.  One is for device operations but the other case is for chaining device operations.  We should never produce tensors on host when chaining device operations in the monad pattern.  See ToMemoryConfig for an example of this chainging pattern.